lib/Remotery.c

```diff
@@ -2347,13 +2365,13 @@ static ObjectLink* ObjectAllocator_Pop(ObjectAllocator* allocator)
     // CAS pop from the front of the list
     for (;;)
     {
-        ObjectLink* old_link = (ObjectLink*)allocator->first_free;
+        rmtAtomicPtr old_link = allocator->first_free;
         if (old_link == NULL)
         {
             return NULL;
         }
-        ObjectLink* next_link = old_link->next;
-        if (AtomicCompareAndSwapPointer((long* volatile*)&allocator->first_free, (long*)old_link, (long*)next_link) ==
+        rmtAtomicPtr next_link = ((ObjectLink*) (old_link))->next;
+        if (AtomicCompareAndSwapPointer(&old_link, old_link, next_link) ==
             RMT_TRUE)
         {
             link = old_link;
```

Comment:
I tried to follow the same logic as the previous code.

In the previous example, line 2333 performs an unprotected read which would trigger a warning:
```ObjectLink* old_link = (ObjectLink*)allocator->first_free;```

Then line 2335 casts the first argument into a `volatile` one:
```if (AtomicCompareAndSwapPointer((long* volatile*)&allocator->first_free, (long*)old_link, (long*)start) ==```

that will be used in:
```static rmtBool AtomicCompareAndSwapPointer(long* volatile* ptr, long* old_ptr, long* new_ptr)```

For the code I proposed, I changed those types to proper Atomic types.

In my understanding, this would require to change the type of the field of the structures (e.g. ``struct ObjectLink_s* volatile next;``) to a `rmtAtomicPtr*`. However, I decided to try the previous approach of first casting it to a pseudo atomic (i.e. `volatile`, instead this time it's a proper atomic type). This removed the warnings, but I'm not sure of the semantics.

A better approach might be to change the type of the fields for which we want to protect read and write with multiple threads?

Reply:
But, at the point of AtomicCompareAndSwapPointer, the `allocator->first_free` may have changed, and you're passing in the address of a pointer allocated on the stack? It looks strange to me.